File: rust/src/qmp.rs
//! QMP (QEMU Monitor Protocol) client.
//!
//! Connects to a QEMU QMP Unix socket, negotiates capabilities,
//! sends a JSON command, and returns the response.  Host-side
//! native Rust implementation callable directly from Python via
//! `PyO3`.  Guest-side command execution uses QGA (QEMU Guest Agent)
//! via `vm.py`, not this module.

use std::io::{self, BufRead, BufReader, Write};
use std::os::unix::net::UnixStream;
use std::time::Duration;

/// Send a QMP command and return the JSON response string.
///
/// Protocol flow:
/// 1. Connect to Unix socket
/// 2. Read greeting (contains `"QMP"` key)
/// 3. Send `{"execute":"qmp_capabilities"}`
/// 4. Read capabilities response
/// 5. Send the user command
/// 6. Skip async events, return the first line containing `"return"` or `"error"`
pub fn qmp_send(socket_path: &str, command_json: &str, timeout_secs: u64) -> io::Result<String> {
    let stream = UnixStream::connect(socket_path)?;
    let timeout = Duration::from_secs(timeout_secs);
    stream.set_read_timeout(Some(timeout))?;
    stream.set_write_timeout(Some(timeout))?;

    let mut reader = BufReader::new(stream.try_clone()?);
    let mut writer = stream;

    // 1. Read QMP greeting
    let mut line = String::new();
    reader.read_line(&mut line)?;
    // Greeting should contain "QMP"
    if !line.contains("\"QMP\"") {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("unexpected QMP greeting: {}", line.trim()),
        ));
    }

    // 2. Send qmp_capabilities
    writer.write_all(b"{\"execute\":\"qmp_capabilities\"}\n")?;
    writer.flush()?;

    // 3. Read capabilities response
    line.clear();
    reader.read_line(&mut line)?;
    // Should contain "return"
    if !line.contains("\"return\"") {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("qmp_capabilities failed: {}", line.trim()),
        ));
    }

    // 4. Send user command
    writer.write_all(command_json.as_bytes())?;
    writer.write_all(b"\n")?;
    writer.flush()?;

    // 5. Read response, skip async events
    loop {
        line.clear();
        let n = reader.read_line(&mut line)?;
        if n == 0 {
            return Err(io::Error::new(
                io::ErrorKind::UnexpectedEof,
                "QMP connection closed before response",
            ));
        }
        let trimmed = line.trim();
        if trimmed.contains("\"return\"") || trimmed.contains("\"error\"") {
            return Ok(trimmed.to_string());
        }
        // else: async event (e.g. RESUME), skip
    }
}

File: go/internal/qmp/qmp.go
// Package qmp implements the QEMU Monitor Protocol client.
package qmp

import (
	"bufio"
	"fmt"
	"io"
	"net"
	"strings"
	"time"
)

// Send connects to a QMP Unix socket, negotiates capabilities,
// sends a JSON command, and returns the response.
func Send(socketPath, commandJSON string, timeoutSecs uint64) (string, error) {
	conn, err := net.Dial("unix", socketPath)
	if err != nil {
		return "", err
	}
	defer conn.Close()

	timeout := time.Duration(timeoutSecs) * time.Second
	_ = conn.SetDeadline(time.Now().Add(timeout))

	reader := bufio.NewReader(conn)

	// 1. Read QMP greeting
	line, err := reader.ReadString('\n')
	if err != nil {
		return "", err
	}
	if !strings.Contains(line, `"QMP"`) {
		return "", fmt.Errorf("unexpected QMP greeting: %s", strings.TrimSpace(line))
	}

	// 2. Send qmp_capabilities
	if _, err := io.WriteString(conn, "{\"execute\":\"qmp_capabilities\"}\n"); err != nil {
		return "", err
	}

	// 3. Read capabilities response
	line, err = reader.ReadString('\n')
	if err != nil {
		return "", err
	}
	if !strings.Contains(line, `"return"`) {
		return "", fmt.Errorf("qmp_capabilities failed: %s", strings.TrimSpace(line))
	}

	// 4. Send user command
	if _, err := io.WriteString(conn, commandJSON+"\n"); err != nil {
		return "", err
	}

	// 5. Read response, skip async events
	for {
		line, err = reader.ReadString('\n')
		if err != nil {
			return "", fmt.Errorf("QMP connection closed before response")
		}
		trimmed := strings.TrimSpace(line)
		if strings.Contains(trimmed, `"return"`) || strings.Contains(trimmed, `"error"`) {
			return trimmed, nil
		}
	}
}
